src/query/functions/temporal/base.go

```diff
@@ -119,191 +128,204 @@ func (c *baseNode) Process(
-	// cache the current block.
-	if emptyLeftBlocks || emptyRightBlocks {
-		if err := c.cache.add(bounds.Start, unconsolidatedBlock); err != nil {
-			return err
-		}
+	m := blockMeta{
+		end:         bounds.Start.UnixNano(),
+		seriesMeta:  resultSeriesMeta,
+		aggDuration: int64(c.op.duration),
+		stepSize:    int64(bounds.StepSize),
+		steps:       steps,
 	}
 
-	blocks, err := c.processCompletedBlocks(queryCtx, processRequests, maxBlocks)
+	concurrency := runtime.NumCPU()
+	batches, err := unconsolidatedBlock.MultiSeriesIter(concurrency)
 	if err != nil {
-		return err
+		// NB: If the unconsolidated block does not support multi series iteration,
+		// fallback to processing series one by one.
+		singleProcess(seriesIter, builder, m, c.processor)
+	} else {
+		batchProcess(batches, builder, m, c.processor)
 	}
 
-	defer closeBlocks(blocks)
+	// NB: safe to close the block here.
+	if err := b.Close(); err != nil {
+		return err
+	}
 
-	return c.propagateNextBlocks(processRequests, blocks, maxBlocks)
+	bl := builder.Build()
+	defer bl.Close()
+	return c.controller.Process(queryCtx, bl)
 }
 
-func closeBlocks(blocks []block.Block) {
-	for _, bl := range blocks {
-		bl.Close()
-	}
+type blockMeta struct {
+	end         int64
+	aggDuration int64
+	stepSize    int64
+	steps       int
+	seriesMeta  []block.SeriesMeta
 }
 
-// processCurrent processes the current block. For the current block,
-// figure out whether we have enough previous blocks which can help process it.
-func (c *baseNode) processCurrent(
-	bounds models.Bounds,
-	leftRangeStart models.Bounds,
-) ([]block.UnconsolidatedBlock, bool, error) {
-	numBlocks := bounds.Blocks(leftRangeStart.Start)
-	leftBlks, err := c.cache.multiGet(leftRangeStart, numBlocks, true)
-	if err != nil {
-		return nil, false, err
-	}
-	return leftBlks, len(leftBlks) != numBlocks, nil
-}
+func batchProcess(
+	iterBatches []block.UnconsolidatedSeriesIterBatch,
+	builder block.Builder,
+	m blockMeta,
+	p processor,
+) error {
+	var (
+		metas = m.seriesMeta
 
-// processRight processes blocks after current block. This is done by fetching
-// all contiguous right blocks until the right range.
-func (c *baseNode) processRight(
-	bounds models.Bounds,
-	rightRangeStart models.Bounds,
-) ([]block.UnconsolidatedBlock, bool, error) {
-	numBlocks := rightRangeStart.Blocks(bounds.Start)
-	rightBlks, err := c.cache.multiGet(bounds.Next(1), numBlocks, false)
-	if err != nil {
-		return nil, false, err
-	}
+		mu       sync.Mutex
+		wg       sync.WaitGroup
+		multiErr xerrors.MultiError
+		idx      int
+	)
 
-	return rightBlks, len(rightBlks) != numBlocks, nil
+	builder.PopulateColumns(len(metas))
+	for _, batch := range iterBatches {
+		wg.Add(1)
+		// capture loop variables
+		loopIndex := idx
+		batch := batch
+		idx = idx + batch.Size
+		go func() {
+			err := buildBlockBatch(loopIndex, batch.Iter, builder, m, p, &mu)
+			mu.Lock()
+			// NB: this no-ops if the error is nil.
```

Comment:
We could skip the `mu.Lock()` however if nil. Is that worth it?

Reply:
Sounds good, may handle that in the cleanup PR if thats ok